doc/ldap2csv.1.in
.\"
.\" LDAP Utilities
.\" Copyright (C) 2019 David M. Syzdek <david@syzdek.net>.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"    1. Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"    2. Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in the
.\"       documentation and/or other materials provided with the distribution.
.\"
.\"    3. Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived from
.\"       this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
.\" IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
.\" THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" doc/ldap2csv.1.in - man page for ldap2csv
.\"
.TH "LDAP2CSV" "1" "@RELEASE_MONTH@" "@PACKAGE_NAME@" "User Commands"
.SH NAME
ldap2csv \- LDAP search tool which outputs CSV


.SH SYNOPSIS
\fBldap2csv\fR
[\fB-b\fR \fIbasedn\fR]
[\fB-c\fR]
[\fB-d\fR \fIlevel\fR]
[\fB-D\fR \fIbinddn\fR]
[\fB-H\fR \fIURI\fR]
[\fB-l\fR \fIlimit\fR]
[\fB-L\fR[\fB-L\fR]]
[\fB-n\fR]
[\fB-v\fR | \fB--version\fR]
[\fB-s\fR \fIscope\fR]
[\fB-S\fR \fIattr\fR]
[\fB-w\fR \fIpasswd\fR]
[\fB-W\fR]
[\fB-x\fR]
[\fB-y\fR \fIfile\fR]
[\fB-Y\fR \fImech\fR]
[\fB-z\fR \fIlimit\fR]
[\fB-Z\fR[\fB-Z\fR]]
[\fIfilter\fR]
\fIattributes[:value[:title]] ...\fR
.sp
\fBldap2csv\fR [ \fB-h\fR | \fB--help\fR ]
.sp
\fBldap2csv\fR [ \fB-V\fR | \fB--version\fR ]


.SH DESCRIPTION
ldap2csv is a shell utilty which performs an LDAP search and prints the results
in CSV format. 


.SH OPTIONS
.TP
\fB-c\fR
do not stop if an error is encountered
.TP
\fB-d\fR
set OpenLDAP debug level to `level'
.TP
\fB-D\fR \fIbinddn\fR
bind DN used for simple bind
.TP
\fB-H\fR \fIURI\fR
specifies list of LDAP Uniform Resource Identifier(s) used to connect to LDAP server
.TP
\fB-l\fR \fIlimit\fR
time limit (in seconds) for search
.TP
\fB-L\fR
two \fB-L\fR disables comments
.TP
\fB-v\fR   \fB--version\fR
run in verbose mode
.TP
\fB-s\fR \fIscope\fR
specifies search filter. Must be one of \fIbase\fR, \fIone\fR, \fIsub\fR, or \fIchild\fR
.TP
\fB-S\fR \fIattr\fR
sort results by attribute \fIattr\fR
.TP
\fB-w\fR \fIpasswd\fR
bind password used for simple bind
.TP
\fB-W\fR
prompt for bind password used in simple bind
.TP
\fB-x\fR
use simple authentication for bind
.TP
\fB-y\fR \fIfile\fR
read bind password from file
.TP
\fB-Y\fR \fImech\fR
SASL mechanism used during bind
.TP
\fB-z\fR \fIlimit\fR
size limit for search
.TP
\fB-Z\fR[\fB-Z\fR]
Issue  StartTLS before bind request. \fB-ZZ\fR requires TLS operations to be successful. 
.TP
\fIfilter\fR
The search filter. If not provided, the default filter, \fB(objectclass=*)\fR,
is used.
.TP
\fIattribute\fR
The \fIattribute\fR to include in CSV output.  The psuedo attributes
\fBdn\fR, \fBrdn\fR, \fBufn\fR, \fBadc\fR, and \fBdce\fR are supported.
.TP
\fIattribute:value\fR
The \fIattribute\fR and default value to include in CSV output. The default
value is displayed if the entry does not contain the specified attribute.
Psuedo attributes cannot be used with default values.
.TP
\fIattribute:value:title\fR
The \fIattribute\fR, default \fIvalue\fR, and \fItitle\fR of column to include
in CSV output. The default value is displayed if the entry does not contain
the specified attribute. Psuedo attributes cannot be used with default values.
.TP
\fI...\fR
List of additional attribute, default values, and titles to include in CSV
output.
.SH PSUEDO ATTRIBUTES
.TP
\fBdn\fR
distinguished name of entry. The following are examples of distinguished name:
.in +4n
.nf

uid=dnullman,ou=People,dc=example,dc=net,o=internet
uid=jdough,ou=People,dc=example,dc=net,o=internet
uid=syzdek,ou=People,dc=syzdek,dc=net,o=internet
uid=administrator,ou=People,dc=foo,dc=org

.fi
.in
.TP
\fBrdn\fR
relative distinguished name of entry. The following are examples of relative
distinguished name:
.in +4n
.nf

uid=dnullman
uid=jdough
uid=syzdek
uid=administrator

.fi
.in
.TP
\fBufn\fR
User Friendly Name of distinguished name. The following are examples of User
Friendly Names:
.in +4n
.nf

dnullman, People, syzdek, net, internet
jdough, People, example, net, internet
syzdek, People, example, net, internet
administrator, People, foo.org

.fi
.in
.TP
\fBadc\fR
Active Directory canonical name of entry. The following are examples of Active
Directory canonical names:
.in +4n
.nf

internet/net/example/People/dnullman/
internet/net/example/People/jdough/
internet/net/syzdek/People/syzdek/
foo.org/People/administrator

.fi
.in
.TP
\fBdce\fR
DCE-style of distinguished name. The following are examples of ADCE-style of
distinguished names:
.in +4n
.nf

/o=internet/dc=net/dc=example/ou=People/uid=dnullman
/o=internet/dc=net/dc=example/ou=People/uid=jdough
/o=internet/dc=net/dc=syzdek/ou=People/uid=syzdek
/dc=org/dc=foo/ou=People/uid=administrator

.fi
.in

.SH EXAMPLE
The following command:
.in +4n
.nf

ldap2csv -LLL -x -b o=internet -S sn '(uid=*)' uid givenname sn mail title rdn

.fi
.in

will perform search on the LDAP server.  The output might look something the following:
.in +4n
.nf

"uid","givenname","sn","mail","title","rdn"
"jdough","John","Dough","doughboy42@example.com","'Dough' Master","uid=jdough"
"dnullman","Devian","Nullman","noreply@example.com","Linux Device : /dev/null","uid=dnullman"
"syzdek","David M.","Syzdek","david@syzdek.net","Slackware Linux Administrator","uid=syzdek"

.fi
.in

Similar search performed with \fBldapsearch\fR might produce output which looks something like the following:

.in +4n
.nf

dn: uid=jdough,ou=People,dc=example,dc=net,o=internet
uid: jdough
givenname: John
sn: Dough 
mail: doughboy42@example.com
title: "Dough" Master

dn: uid=dnullman,ou=People,dc=example,dc=net,o=internet
uid: dnullman
givenname: Devian
sn: Nullman
mail: noreply@example.com
title: Linux Device | /dev/null

dn: uid=syzdek,ou=People,dc=syzdek,dc=net,o=internet
uid: syzdek
givenname: David M.
sn: Syzdek
mail: david@syzdek.net
title: Slackware Linux Administrator

.fi
.in


.SH "SEE ALSO"
.BR ldapsearch (1),
.BR ldapdebug (1),
.BR ldaptree (1),
.BR ldap.conf (5)


.SH AUTHOR
David M. Syzdek <david@syzdek.net>


.SH ACKNOWLEDGEMENTS
\fB@PACKAGE_NAME@\fR is developed and maintained by David M. Syzdek
<david@syzdek.net>. \fB@PACKAGE_NAME@\fR utilizes and is styled after the
tools and libraries maintained by the \fBThe OpenLDAP Project\fR
<http://www.openldap.org/>.


.Sh CAVEATS

.\" end of man page


ChangeLog

LDAP Utilities
Copyright (C) 2012 Bindle Binaries <syzdek@bindlebinaries.com>.

0.3
      - refactoring libldaputils (syzdek)
      - adding TLS support

0.2   
      - Recreating Xcode project file. (syzdek)
      - Removing files installed by autotools from git repo. (syzdek)
      - Updating autotools files (Makefile.am, configure.ac, etc). (syzdek)
      - Changing versioning scheme. (syzdek)
      - Re-writing README. (syzdek)

0.1.1
      - Updating ldaputils_getpass() to use stderr.
      - Modifying behavior of --verbose to set debug level
      - Removing debugging lines for LDAP API Version.
      - Updating version information to include LDAP API information.
      - Adding sorting functions for LDAP searches.
      - Fixing problems compiling with MinGW.
      - Abstracting code for separating multivalue values.
      - Fixing gettext international support.
      - Changing how LDAP host/port/URI values are stored so that values
        from ldap.conf and ldaprc files will be used if they are not
        specified on the command line. (syzdek)
      - Changing default search scope to SUBTREE. (syzdek)
      - Changing version major version from 1.0.0. to 0.1.0 (syzdek)

0.1.0   2008/10/12
	* initial release of package (syzdek)


src/ldap2csv.c
/*
 *  LDAP Utilities
 *  Copyright (C) 2012, 2019 David M. Syzdek <david@syzdek.net>.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are
 *  met:
 *
 *     1. Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *
 *     2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *
 *     3. Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived from
 *        this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 *  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 *  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/**
 *  @file src/ldap2csv.c export LDAP data to CSV file
 */
/*
 *  Simple Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldap2csv" -Wall -I../include'
 *     gcc ${CFLAGS} -c ldap2csv.c
 *     gcc ${CFLAGS} -lldap -o ldap2csv ldap2csv.o ../lib/libldaputils.a
 *
 *  Libtool Build:
 *     export CFLAGS='-DPROGRAM_NAME="ldap2csv" -Wall -I../include'
 *     libtool --mode=compile --tag=CC gcc ${CFLAGS} -c ldap2csv.c
 *     libtool --mode=link    --tag=CC gcc ${CFLAGS} -lldap -o ldap2csv \
 *             ldap2csv.lo ../lib/libldaputils.a
 *
 *  Libtool Clean:
 *     libtool --mode=clean rm -f ldap2csv.lo ldap2csv
 */
#define _LDAP_UTILS_SRC_LDAP2CSV 1
#undef __LDAPUTILS_PMARK


///////////////
//           //
//  Headers  //
//           //
///////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Headers
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <getopt.h>
#include <assert.h>

#define LDAP_DEPRECATED 1
#include <ldap.h>
#include <ldaputils.h>
#include <ldapschema.h>


///////////////////
//               //
//  Definitions  //
//               //
///////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Definitions
#endif

#ifndef PROGRAM_NAME
#define PROGRAM_NAME "ldap2csv"
#endif

#define MY_SHORT_OPTIONS LDAPUTILS_OPTIONS_COMMON LDAPUTILS_OPTIONS_SEARCH "o:"


/////////////////
//             //
//  Datatypes  //
//             //
/////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Datatypes
#endif

/* configuration union */
typedef struct my_config MyConfig;
struct my_config
{
   LDAPUtils   * lud;
   LDAPSchema  * lsd;
   const char  * filter;
   const char  * prog_name;
   const char ** defvals;
   const char ** titles;
   char          output[LDAPUTILS_OPT_LEN];
};


//////////////////
//              //
//  Prototypes  //
//              //
//////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Prototypes
#endif

// main statement
int main(int argc, char * argv[]);

// parses configuration
int my_config(int argc, char * argv[], MyConfig ** cnfp);

int my_results(MyConfig * cnf, LDAPMessage * res);

// fress resources
void my_unbind(MyConfig * cnf);


/////////////////
//             //
//  Functions  //
//             //
/////////////////
#ifdef __LDAPUTILS_PMARK
#pragma mark - Functions
#endif

/// prints program usage and exits
void ldaputils_usage(void)
{
   printf("Usage: %s [options] [filter] attributes[:values[:title]]...\n", PROGRAM_NAME);
   ldaputils_usage_search(MY_SHORT_OPTIONS);
   ldaputils_usage_common(MY_SHORT_OPTIONS);
   printf("Special Attributes:\n");
   printf("  dn                        entry's DN\n");
   printf("  rdn                       entry's relative DN\n");
   printf("  ufn                       entry's User Friendly Name\n");
   printf("  adc                       entry's Active Directory canonical name\n");
   printf("  dce                       entry's DN in DCE-style\n");
   printf("\nReport bugs to <%s>.\n", PACKAGE_BUGREPORT);
   return;
}


/// main statement
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
int main(int argc, char * argv[])
{
   int                    x;
   int                    err;
   MyConfig             * cnf;
   LDAPMessage          * res;

   cnf = NULL;

   // initializes resources and parses CLI arguments
   if ((err = my_config(argc, argv, &cnf)) != 0)
      return(1);
   if (!(cnf))
      return(0);

   // starts TLS and binds to LDAP
   if ((err = ldaputils_bind_s(cnf->lud)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldap_sasl_bind_s(): %s\n", ldaputils_get_prog_name(cnf->lud), ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // fetches schema
   if ( ((err = ldapschema_fetch(cnf->lsd, cnf->lud->ld)) != LDAP_SUCCESS) && (err != LDAPSCHEMA_SCHEMA_ERROR) )
   {
      fprintf(stderr, "%s: ldapschema_fetch(): %s\n", cnf->lud->prog_name, ldapschema_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // performs LDAP search
   if ((err = ldaputils_search(cnf->lud, &res)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldaputils_search(): %s\n", ldaputils_get_prog_name(cnf->lud), ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // prints attribute names
   printf("\"%s\"", cnf->titles[0]);
   for(x = 1; ((cnf->titles[x])); x++)
      printf(",\"%s\"", cnf->titles[x]);
   printf("\n");

   // prints values
   if ((err = my_results(cnf, res)) != LDAP_SUCCESS)
   {
      my_unbind(cnf);
      return(1);
   };

   ldap_msgfree(res);
   my_unbind(cnf);

   return(0);
}


/// parses configuration
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
/// @param[in] cnfp   reference to configuration pointer
int my_config(int argc, char * argv[], MyConfig ** cnfp)
{
   int        c;
   int        err;
   int        option_index;
   char     * str;
   MyConfig * cnf;
   size_t     len;
   size_t     size;

   static char   short_options[] = MY_SHORT_OPTIONS;
   static struct option long_options[] =
   {
      {"help",          no_argument, 0, 'h'},
      {"verbose",       no_argument, 0, 'v'},
      {"version",       no_argument, 0, 'V'},
      {NULL,            0,           0, 0  }
   };

   // allocates memory for configuration
   if (!(cnf = (MyConfig *) malloc(sizeof(MyConfig))))
   {
      fprintf(stderr, "%s: out of virtual memory\n", PROGRAM_NAME);
      return(1);
   };
   memset(cnf, 0, sizeof(MyConfig));

   // initialize ldap utilities
   if ((err = ldaputils_initialize(&cnf->lud, PROGRAM_NAME)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldaputils_initialize(): %s\n", PROGRAM_NAME, ldap_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // initialize ldap schema
   if ((err = ldapschema_initialize(&cnf->lsd)) != LDAP_SUCCESS)
   {
      fprintf(stderr, "%s: ldapschema_initialize(): %s\n", PROGRAM_NAME, ldapschema_err2string(err));
      my_unbind(cnf);
      return(1);
   };

   // loops through args
   option_index = 0;
   while((c = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1)
   {
      switch(ldaputils_getopt(cnf->lud, c, optarg))
      {
         // shared option exit without error
         case -2:
         my_unbind(cnf);
         return(0);

         // no more arguments
         case -1:
         break;

         // long options toggles
         case 0:
         break;

         // shared option error
         case 1:
         my_unbind(cnf);
         return(1);

         // argument error
         case '?':
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);

         // unknown argument error
         default:
         fprintf(stderr, "%s: unrecognized option `--%c'\n", PROGRAM_NAME, c);
         fprintf(stderr, "Try `%s --help' for more information.\n", PROGRAM_NAME);
         my_unbind(cnf);
         return(1);
      };
   };

   cnf->prog_name = ldaputils_get_prog_name(cnf->lud);

   // checks for required arguments
   if (argc < (optind+1))
   {
      fprintf(stderr, "%s: missing required arguments\n", cnf->prog_name);
      fprintf(stderr, "Try `%s --help' for more information.\n", cnf->prog_name);
      my_unbind(cnf);
      return(1);
   };

   // saves filter
   cnf->lud->filter = "(objectclass=*)";
   if ((index(argv[optind], '=')) != NULL)
   {
      cnf->lud->filter = argv[optind];
      optind++;
   };

   // configures LDAP attributes to return in results
   len  = (size_t)(argc - optind) + 1;
   size = sizeof(char *) * len;
   if (!(cnf->lud->attrs = (char **) malloc(size)))
   {
      fprintf(stderr, "%s: out of virtual memory\n", cnf->prog_name);
      my_unbind(cnf);
      return(1);
   };
   bzero(cnf->lud->attrs, size);
   if (!(cnf->defvals = (const char **) malloc(size)))
   {
      fprintf(stderr, "%s: out of virtual memory\n", cnf->prog_name);
      my_unbind(cnf);
      return(1);
   };
   bzero(cnf->defvals, size);
   if (!(cnf->titles = (const char **) malloc(size)))
   {
      fprintf(stderr, "%s: out of virtual memory\n", cnf->prog_name);
      my_unbind(cnf);
      return(1);
   };
   bzero(cnf->titles, size);
   for(c = 0; c < (argc-optind); c++)
   {
      cnf->lud->attrs[c] = argv[optind+c];
      cnf->defvals[c]    = "";
      cnf->titles[c]     = argv[optind+c];
      if ((str = index(argv[optind+c], ':')) != NULL)
      {
         str[0] = '\0';
         cnf->defvals[c] = &str[1];
         if ((str = index(&str[1], ':')) != NULL)
         {
            str[0] = '\0';
            cnf->titles[c] = &str[1];
         };
      };
   };
   cnf->lud->attrs[c] = NULL;
   cnf->defvals[c]    = NULL;
   cnf->titles[c]     = NULL;

   // reads password
   if ((err = ldaputils_pass(cnf->lud)) != 0)
   {
      my_unbind(cnf);
      return(1);
   };

   *cnfp = cnf;

   return(0);
}


// prints results
int my_results(MyConfig * cnf, LDAPMessage * res)
{
   int               x;
   int               y;
   void            * ptr;
   char            * buff;
   size_t            bufflen;
   char            * dn;
   char           ** dns;
   char            * dnstr;
   char            * delim;
   LDAPMessage     * msg;
   struct berval  ** vals;
   LDAP            * ld;
   LDAPSchemaAttributeType * attr;
   char           ** names;

   assert(cnf != NULL);
   assert(res != NULL);

   ld      = ldaputils_get_ld(cnf->lud);

   // initialize buffer
   bufflen = 32;
   if ((buff = malloc(bufflen)) == NULL)
   {
      fprintf(stderr, "%s: malloc(): out of virtual memory\n", cnf->prog_name);
      return(LDAP_NO_MEMORY);
   };

   // sorts entries
   if ((cnf->lud->sortattr))
      ldap_sort_entries(ld, &res, cnf->lud->sortattr, strcasecmp);

   // loops through entries
   msg = ldap_first_entry(ld, res);
   while ((msg))
   {
      printf("\"");

      // retrieve DN and make CSV safe
      if ((dn = ldap_get_dn(ld, msg)) == NULL)
      {
         fprintf(stderr, "%s: malloc(): out of virtual memory\n", cnf->prog_name);
         free(buff);
         return(LDAP_NO_MEMORY);
      };
      delim = dn;
      while((delim = index(delim, '"')) != NULL)
         delim[0] = '\'';

      // loop through attributes
      for(x = 0; (cnf->lud->attrs[x] != NULL); x++)
      {
         // print delimiter
         if (x > 0)
            printf("\",\"");

         // prints dn if specified
         if (strcasecmp("dn", cnf->lud->attrs[x]) == 0)
         {
            printf("%s", dn);
            continue;
         };

         // print RDN
         if (strcasecmp("rdn", cnf->lud->attrs[x]) == 0)
         {
            if ((dns = ldap_explode_dn(dn, 0)) == NULL)
            {
               fprintf(stderr, "%s: ldap_explode_dn(): out of virtual memory\n", cnf->prog_name);
               free(buff);
               return(LDAP_NO_MEMORY);
            };
            printf("%s", dns[0]);
            ldap_value_free(dns);
            continue;
         };

         // print DN in UFN format
         if (strcasecmp("ufn", cnf->lud->attrs[x]) == 0)
         {
            if ((dnstr = ldap_dn2ufn(dn)) == NULL)
            {
               fprintf(stderr, "%s: ldap_dn2ufn(): out of virtual memory\n", cnf->prog_name);
               free(buff);
               return(LDAP_NO_MEMORY);
            };
            printf("%s", dnstr);
            ldap_memfree(dnstr);
            continue;
         };

         // print DN in DCE format
         if (strcasecmp("dce", cnf->lud->attrs[x]) == 0)
         {
            if ((dnstr = ldap_dn2dcedn(dn)) == NULL)
            {
               fprintf(stderr, "%s: ldap_dn2dcedn(): out of virtual memory\n", cnf->prog_name);
               free(buff);
               return(LDAP_NO_MEMORY);
            };
            printf("%s", dnstr);
            ldap_memfree(dnstr);
            continue;
         };

         // print DN in AD canonical format
         if (strcasecmp("adc", cnf->lud->attrs[x]) == 0)
         {
            if ((dnstr = ldap_dn2ad_canonical(dn)) == NULL)
            {
               fprintf(stderr, "%s: ldap_dn2ad_canonical(): out of virtual memory\n", cnf->prog_name);
               free(buff);
               return(LDAP_NO_MEMORY);
            };
            printf("%s", dnstr);
            ldap_memfree(dnstr);
            continue;
         };

         // retrieves values
         names = NULL;
         vals  = NULL;
         if ((attr = ldapschema_find_attributetype(cnf->lsd, cnf->lud->attrs[x])) != NULL)
            ldapschema_get_info_attributetype(cnf->lsd, attr, LDAPSCHEMA_FLD_NAME, &names);
         for(y = 0; (((names)) && ((names[y])) && (!(vals))); y++)
            vals = ldap_get_values_len(ld, msg, names[y]);
         if (!(vals))
            vals = ldap_get_values_len(ld, msg, cnf->lud->attrs[x]);
         if (!(vals))
         {
            printf("%s", cnf->defvals[x]);
            continue;
         };

         // processes values
         for(y = 0; (y < ldap_count_values_len(vals)); y++)
         {
            // adjusts size of buffer
            if (bufflen < (vals[y]->bv_len + 1))
            {
               bufflen = vals[y]->bv_len + 1;
               if ((ptr = realloc(buff, bufflen)) == NULL)
               {
                  fprintf(stderr, "%s: realloc(): out of virtual memory\n", cnf->prog_name);
                  free(buff);
                  return(LDAP_NO_MEMORY);
               };
               buff = ptr;
            };

            // copies value into buffer
            memcpy(buff, vals[y]->bv_val, vals[y]->bv_len);
            buff[vals[y]->bv_len] = '\0';

            // replace double quotation character with single quotation character
            delim = buff;
            while((delim = index(delim, '"')) != NULL)
               delim[0] = '\'';
            delim = buff;
            while((delim = index(delim, '|')) != NULL)
               delim[0] = ':';

            // print value
            if (y > 0)
               printf("|%s", buff);
            else
               printf("%s", buff);
         };
         ldap_value_free_len(vals);
      };
      printf("\"\n");

      // frees DN
      ldap_memfree(dn);

      // retrieves next entry
      msg = ldap_next_entry(ld, msg);
   };

   free(buff);

   return(LDAP_SUCCESS);
}


// fress resources
void my_unbind(MyConfig * cnf)
{
   assert(cnf != NULL);

   if ((cnf->lsd))
      ldapschema_free(cnf->lsd);

   if ((cnf->lud))
      ldaputils_unbind(cnf->lud);

   if ((cnf->defvals))
      free(cnf->defvals);

   if ((cnf->titles))
      free(cnf->titles);

   free(cnf);

   return;
}

/* end of source file */
